compiler/rustc_typeck/src/expr_use_visitor.rs

```diff
@@ -671,6 +677,13 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                         delegate.bind(binding_place, binding_place.hir_id);
                     }
 
+                    // Subtle: MIR desugaring introduces immutable borrows for each pattern
+                    // binding when lowering pattern guards to ensure that the guard does not
+                    // modify the scrutinee.
+                    if has_guard {
```

Comment:
The tests pass with or without this change, so arguably we should remove it. That said, this feels like the more correct behavior, so I think we should leave it in.

Reply:
Hmm. Yes, it seems correct, but it's interesting that the tests pass either way! There is a certain amount of redundancy, it seems, between the analysis of what is borrowed etc and the "what types appear in the generator". 

I think perhaps we could observe the difference with this test, but maybe not? 

I think this will compile successfully:

```rust
#![feature(generators)]

fn is_send<T: Send>(_: T) { }

fn main() {
    is_send(async move {
        let x = std::rc::Rc::new(22);
        //match x { ref y if foo().await => () }
        drop(x);
        foo().await;
    });
}

async fn foo() -> bool {
    true 
}
```

but if you remove the comment from `//match`, it will not. I'm curious what happens if you remove this line here?

I guess that to really eliminate this redundancy we would drive the "do we need to include &T" check based on some set generated by this code (e.g., "is lvalue borrowed" or whatever).

Reply:
The test you provided does compile successfully. When I uncomment the `match` line, it gives a compiler error saying that the async block is not `Send` due to `y` and `x` being borrowed across the await point.

The behavior is the same with and without the `if has_guard { ... }` section in `expr_use_visitor.rs`.